### ISSUE TITLE
`@library` decorator that supports configuring and forces using `@keyword` to mark keywords

### ISSUE THREAD

**Issue by yahman72** [robotframework/robotframework#3019]
I guess this would be a new feature: have a look an let me know what you think about it.

The "Problem":
Sometimes you have python code with class structure like this
```
             BaseClass
             foo()
                |
       ---------------------
       |                   |
   SubClassA           SubClassB
   bar_a()             bar_b()
```

And you would like to make a RF Library out of it like this:
```
             BaseClass
             foo()
                |
       ---------------------
       |                   |
   SubClassA           SubClassB
   bar_a()             bar_b()
       |                   |
       |                   |
   RFLibA              RFLibB
   foo_bar()           bar_foo()
```

But when trying to use those `RFLibA` and `RFLibB` libraries you'll run into to trouble because e.g. `foo()` is in both libraries.

It would be good to be able to hide certain KWs in a library.

I had a look at the code and the `@keyword` decorator looks like nice fit for this i.e. reserve a value for the name argument that is treated as "hide this keyword" - for example the string "_" (could also be something else

I.e. in `RFLibA` I would have this for hiding the `foo()` KW
```
    @keyword(name="_")
    def foo(self):
        super(RFLibA).foo()
```
This could be handled e.g. in the `robot.running.testlibraries._BaseTestLibrary` e.g in the `_create_handlers()`

I could also do the same (i.e. have `RFLibA` without `foo()`) without this "hide" feature also by creating the libraries in a different way:
- `RFLibA` is a wrapper that is not inherited from `SubClassA`, but creates an instance of `SubClassA` and calls it's methods
- override `__dir__()` in `RFLibA` and explicitly list the KWs I want to expose --> a bit of a hack
 
But in my current case (much more complex class hiearchy) neither of this is desirable - that "hide" would be the nicest & cleanest way.

**Comment by pekkaklarck** [comment]
First of all, I highly recommend avoiding inheritance hierarchies like this and using composition instead. You already mentioned that `RFLibA` doesn't necessarily need to inherit `SubClassA` but instead it could use it internally.

There are, however, also other reasons to expose only certain methods of a class as keywords. Using the `@keyword` decorator to mark which methods should be considered keywords is a good idea. For backwards compatibility reasons we cannot change the library API so that others would be ignored automatically, though.

I don't like the idea that non-keyword methods needed to be marked with `@keyword(name='_')` or even with more explicit decorator like `@not_keyword`. This wouldn't even work in your case if you didn't control `SubClassA` code and couldn't add these decorators there. I propose an alternative idea that I've been already thinking earlier in a separate comment.

**Comment by pekkaklarck** [comment]
In my opinion the best way to limit what methods become keywords would be adding a new class decorator `@library` that could be used with the existing `@keyword` decorator like this:

```python
from robot.api.deco import keyword, library

@library
class Example:

    @keyword
    def this_is_keyword(self):
        # ...

    def this_is_not_keyword(self):
        # ...
```

The idea is that if Robot detects that a library has been decorated with `@library`, only methods decorated with `@keyword` are considered keyword. Basically the `@library` decorator just needed to add some special attribute to the library class (e.g. `ROBOT_LIBRARY = True`) that Robot can then check. Shouldn't be too complicated to implement and would be a nice addition to RF 3.2 if someone is interested to implement it!

Note:

- You can actually already now use [PythonLibCore](https://github.com/robotframework/PythonLibCore/) to accomplish more or less what the `@library` decorator would do. It has also other use cases, and having this functionality built-in would be nice.

- The new decorator could also get new functionality like `@library(scope=library.GLOBAL, version='0.1a1')` afterwards.

**Comment by yahman72** [comment]
That `@Library` sounds indeed better and I'll definitely have a look at the `pythonlibcore`, thanks!

If (when) my "pain" with that library of mine gets too big I would be interested in implementing this, but this is most likely not going to happen within the next few months.

**Comment by pekkaklarck** [comment]
This was implemented by @cristii006 in PR #3222. I've enhanced tests and docs a bit in the above commits. Still need to enhanced the documentation in the User Guide but otherwise this is ready.
